file/file_util.cc

```diff
@@ -123,22 +123,33 @@ bool IsWalDirSameAsDBPath(const ImmutableDBOptions* db_options) {
   return same;
 }
 
-IOStatus GenerateOneFileChecksum(FileSystem* fs, const std::string& file_path,
-                                 FileChecksumGenFactory* checksum_factory,
-                                 std::string* file_checksum,
-                                 std::string* file_checksum_func_name,
-                                 size_t verify_checksums_readahead_size,
-                                 bool allow_mmap_reads,
-                                 std::shared_ptr<IOTracer>& io_tracer) {
+// requested_checksum_func_name brings the function name of the checksum
+// generator in checksum_factory. Checksum factories may use or ignore
+// requested_checksum_func_name.
+IOStatus GenerateOneFileChecksum(
+    FileSystem* fs, const std::string& file_path,
+    FileChecksumGenFactory* checksum_factory,
+    const std::string& requested_checksum_func_name, std::string* file_checksum,
+    std::string* file_checksum_func_name,
+    size_t verify_checksums_readahead_size, bool allow_mmap_reads,
+    std::shared_ptr<IOTracer>& io_tracer) {
   if (checksum_factory == nullptr) {
     return IOStatus::InvalidArgument("Checksum factory is invalid");
   }
   assert(file_checksum != nullptr);
   assert(file_checksum_func_name != nullptr);
 
   FileChecksumGenContext gen_context;
+  gen_context.requested_checksum_func_name = requested_checksum_func_name;
   std::unique_ptr<FileChecksumGenerator> checksum_generator =
       checksum_factory->CreateFileChecksumGenerator(gen_context);
+  if (checksum_generator == nullptr) {
```

Comment:
QQ: what does it mean if `checksum_generator != nullptr && checksum_generator->Name() != requested_checksum_func_name`?

Reply:
This should not happened, but yes, this behavior undefined. In regular case, the developer of factory should return the checksum_generator->Name() matches with requested_checksum_func_name.

Reply:
Maybe use assertion explicitly disallow this case in debug mode?

Reply:
Sure

Reply:
@riversand963 I think, due to the backward compatibility issue, we cannot do assert here. Before `requested_checksum_func_name` was introduced, we do not give the name to `CreateFileChecksumGenerator`. So checksum_generator->Name() will returns the name and requested_checksum_func_name is empty, it is expected.

Reply:
How about this?
```
assert(!checksum_generator || requested_checksum_func_name.empty() || requested_checksum_func_name == checksum_generator->Name());
```